rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,397 @@
+This situation can disrupt the cluster, as all incoming admission requests will be denied in the absence of operational Policy Server instances.
+
+The proposed changes aim to address these issues by introducing a new policy lifecycle that includes the following features:
+
+- Hot reload of policies: The Policy Server will be able to load a new policy without requiring a restart.
+- Policy validation: The Policy Server will validate the policy before loading it, ensuring that it is correctly compiled and can be loaded.
+- Policy status monitoring: The Policy Server will report the status of the policy, indicating whether it was loaded successfully or if an error occurred.
+- Policy versioning: The Policy Server will keep running with a previous version of the policy in case of issues with the latest policy.
+- Policy rollback: The user will be able to rollback to a previous policy version in case of issues with the latest policy.
+
+## Examples / User Stories
+
+[examples]: #examples
+
+- As a user, I want to load a new policy/update an existing policy without requiring a restart of the policy server.
+- As a user, I want to know if a policy was pulled or validated successfully.
+- As a user, I want to monitor the status of a policy on specific policy server replica so that I can identify and resolve issues with the policy.
+- As a user, I want that the policy server keeps running with a previous version of the policy in case of issues with the latest policy.
+- As a user, I want to rollback to a previous policy version in case of issues with the latest policy.
+- As a user, I want to debug issues with the namespaced policies I own without needing to involve the cluster administrator.
+
+# Detailed design
+
+[design]: #detailed-design
+
+> Note: In this RFC, the term `Policy` will be used to refer to any of the following resources: `AdmissionPolicy`, `ClusterAdmissionPolicy`, `AdmissionPolicyGroup`, or `ClusterAdmissionPolicyGroup`.
+
+## Concepts
+
+To implemenet the new policy lifecycle, we will introduce the following concepts:
+
+### Hot reload
+
+The Policy Server will be able to load a new policy without requiring a restart.
+This is will be achieved by refactoring the `EvaluationEnvironment` adding methods to set, update and remove policies.
+
+### The Policy Server acts as a controller
+
+At the time of writing the Policy Server is configured with a ConfigMap that contains all the policies to be loaded.
+When a new policy is created, the ConfigMap is updated, and the Policy Server is restarted to load the new policy.
+
+We want to change this behavior to allow the Policy Server to act as a controller that can load new policies without requiring a restart.
+
+#### PolicyRevision CRD
+
+We will introduce a new CRD called `PolicyRevision` to store the state of a policy in relation to the Policy Server replicas.
+`PolicyRevisions` are namespaced resources, and they will be created in the same namespace as the controller.
```

Comment:
by controller, we mean the policy-server that serves the policies associated to the PolicyRevision, correct?

Reply:
we mean the main controller, the kw controller

Reply:
I will clarify